pkg/ignore/ignore_test.go

```diff
@@ -0,0 +1,227 @@
+	for _,fileToKeep :=range filesToKeep {
+		filesToKeepCheck[fileToKeep] = fileToKeep
+		filesToCreate = append(filesToCreate, fileToKeep)
+	}
+
+	
+	// create files for test
+        for _, fileToCreate := range filesToCreate {
+		fbpath := filepath.Join(dpath, fileToCreate)
+
+		// ensure any subdirs off working dir exist
+		dirpath := filepath.Dir(fbpath)
+		derr := os.MkdirAll(dirpath,0777)
+		if derr != nil && !os.IsExist(derr) {
+			t.Errorf("Problem creating subdirs %s with %v \n", dirpath, derr)
+		}
+		t.Logf("Going to create file %s given supplied suffix %s \n", fbpath, fileToCreate)
+		fbfile, fberr := os.Create(fbpath)
+		defer fbfile.Close()
+		if fberr != nil {
+			t.Errorf("Problem creating test file %v \n", fberr)
+		}
+        }
+
+
+	// run ignorer algorithm
+	ignorer := &DockerIgnorer{}	
+	ignorer.Ignore(c)
+
+	// check if filesToDel, minus ignores, are gone, and filesToKeep are still there
+	for _,fileToCheck := range filesToCreate {
+		fbpath := filepath.Join(dpath, fileToCheck)
+		t.Logf("Evaluating file %s from dir %s and file to check %s \n", fbpath, dpath, fileToCheck)
+
+		// see if file still exists or not
+		ofile, oerr := os.Open(fbpath)
+		defer ofile.Close()
+		var fileExists bool
+		if oerr == nil {
+			fileExists = true
+			t.Logf("The file %s exists after Ignore was run \n", fbpath)
+		} else {
+			if os.IsNotExist(oerr) {
+				t.Logf("The file %s does not exist after Ignore was run \n", fbpath)
+				fileExists = false
+			} else {
+				t.Errorf("Could not verify existence of %s because of %v \n", fbpath, oerr)
+			}
+		}
+
+		// if file present, verify it is in ignore or keep list
+		if fileExists {
+			// exclude / ignore
+			_, iok := filesToIgnore[fileToCheck]
+			if iok {
+				t.Logf("validated ignored file is still present %s \n ", fileToCheck)
+				continue
+			}
+
+			// simply specified as file to keep
+			_, kok := filesToKeepCheck[fileToCheck]
+			if kok {
+				t.Logf("validated file to keep is still present %s \n", fileToCheck)
+				continue
+			}
+
+			// if here, then file should not exist
+			t.Errorf("file which should be deleted exists %s \n", fileToCheck)
+			
+		} else {
+			_, dok := filesToDelCheck[fileToCheck]
+			if dok {
+				t.Logf("file which should have been deleted is  gone %s \n", fileToCheck)
+				continue
+			}
+
+			// if here, then file which should still exist has been deleted
+			t.Errorf("file which should exist has been deleted %s \n", fileToCheck)
+			
+		}
+		
+	}
+
+	// clean up test
+	cleanerr := os.RemoveAll(workingDir)
+	if cleanerr != nil {
+		t.Errorf("problem cleaning up %v \n", cleanerr)
+	}
+}
+
+func TestSingleIgnore(t *testing.T) {
+	baseTest(t, []string{"foo.bar\n"}, []string{"foo.bar"}, []string{})
+}
+
+func TestWildcardIgnore(t *testing.T) {
+	baseTest(t, []string{"foo.*\n"}, []string{"foo.a", "foo.b"}, []string{})
+}
+
+func TestExclusion(t *testing.T) {
+	baseTest(t, []string{"foo.*\n", "!foo.a"}, []string{"foo.a","foo.b"}, []string{})
```

Comment:
foo.* occurs first in .s2iignore, and removes foo.a and foo.b.  However, !foo.a follows foo.*, and so excludes that ignore/removal, and so it is kept.